### PR TITLE
Add snowplow tracking for search queries

### LINKED ISSUE #31433: Snowplow: Instrumentation for search performance

### PR DESCRIPTION
**Pull request by calherries**
Resolves https://github.com/metabase/metabase/issues/31433

[Analytics spec](https://www.notion.so/metabase/New-search-query-b92754a13d6f44079bb79da4d18ee6d3?pvs=4)

### CODE REVIEW

File: src/metabase/api/search.clj [hunk 1]
```diff
@@ -515,12 +516,19 @@
...
+  (let [start-time (System/currentTimeMillis)
+        results    (search (search-context
+                            q
+                            archived
+                            table_db_id
+                            models
+                            mw.offset-paging/*limit*
+                            mw.offset-paging/*offset*))
+        duration   (- (System/currentTimeMillis) start-time)]
+    (when (and (nil? models)
```
Comment (qnkhuat) [inline src/metabase/api/search.clj:528]: it's not clear to me why we want this guard, could you explain? and maybe put it into the code as well

File: test/metabase/api/search_test.clj [hunk 2]
```diff
@@ -724,3 +725,12 @@
         ;; we have this test here just to keep tracks this number to remind us to put effort
         ;; into keep this number as low as we can
         (is (= 7 (call-count)))))))
+
+(deftest snowplow-new-search-query-event-test
+  (testing "Send a snowplow event when a new search query is made"
```
Comment (qnkhuat) [inline test/metabase/api/search_test.clj:730]: should we include a test "do not send a snowplow event if archived=true, nil? models, nil? table_id"?

File: src/metabase/api/search.clj (hunk 1 above)
Comment (calherries) [reply src/metabase/api/search.clj:528]: The spec says to track "global search". Any of these arguments limit search in a way that is not global.

Comment (calherries) [reply src/metabase/api/search.clj:529]: Added a comment [Test for only tracking global searches](https://github.com/metabase/metabase/pull/31467/commits/fb06f1ba33bbad2b0dec194f876aaddf224c6330)

File: snowplow/iglu-client-embedded/schemas/com.metabase/search/jsonschema/1-0-0 [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  },
+  "type": "object",
+  "properties": {
+    "event": {
+      "description": "Event name",
+      "type": "string",
+      "enum": ["new_search_query"],
+      "maxLength": 1024
+    },
+    "runtime": {
```
[the remark this answers is not in the document]
Comment (calherries) [reply snowplow/iglu-client-embedded/schemas/com.metabase/search/jsonschema/1-0-0]: Updated

File: test/metabase/api/search_test.clj (hunk 2 above)
Comment (calherries) [reply test/metabase/api/search_test.clj]: Added tests [Test for only tracking global searches](https://github.com/metabase/metabase/pull/31467/commits/fb06f1ba33bbad2b0dec194f876aaddf224c6330)

File: src/metabase/api/search.clj [hunk 4]
```diff
@@ -515,12 +516,20 @@
...
+                            table_db_id
+                            models
+                            mw.offset-paging/*limit*
+                            mw.offset-paging/*offset*))
+        duration   (- (System/currentTimeMillis) start-time)]
+    ;; Only track global searches
+    (when (and (nil? models)
+               (nil? table_db_id)
+               (not archived))
+      (snowplow/track-event! ::snowplow/new-search-query api/*current-user-id* {:runtime-milliseconds duration}))
```
Comment (qnkhuat) [inline src/metabase/api/search.clj:532]:
```suggestion
      (snowplow/track-event! ::snowplow/new-global-search-query api/*current-user-id* {:runtime-milliseconds duration}))
```

should we make it explicit?

Comment (calherries) [reply src/metabase/api/search.clj:532]: I don't think its necessary unless there's more than one snowplow event that needs tracking

**Review by qnkhuat — APPROVED** [review]
LGTM

**Pull request merged**